### PR TITLE
[OPS-7323] File API consolidation and permanent URLs

### PR DESCRIPTION
**Pull request by orakili**
Ticket: [OPS-7323]

This PR is alternative/enhancement of https://github.com/UN-OCHA/docstore-site/pull/93 to have permanent file urls.

In addition to consolidated files API endpoints, this proposes a new approach based on symlinks to control access to files.

**Benefits:**

- Permanent file URL
- Possibility for a provider to select which version of a file to serve (for example until the new version if validated by editors)
- No need to hit the docstore drupal site to serve private files.

## File resources

File resources are *media entities*. A media can have revisions, with each revision pointing at a different *file entity*.

File resources have a permanent URL regardless of their private status: `/files/{uuid}/{filename}`.

Files are store on disk in public://files or private://files with the pattern: `/ab/cd/abcdefgh-abcd-abcd-abcd-abcdefghijkl.ext`

For each media there can be a set of symlinks:

**For public files**

1. Link to latest version

  Pattern: `public://media/latest/ab/cd/{media-uuid}.ext`

  Target: file corresponding to the latest revision of the media

2. Link to a specific version for a provider

  Pattern: `public://media/{provider-uuid}/ab/cd/{media-uuid}.ext`

  Target: file corresponding to a specific revision of the media

3. Link to hide the media for a provider

  Pattern: `public://media/{provider-uuid}/hidden/ab/cd/{media-uuid}.ext`

  Target: file corresponding to the latest revision of the media but it doesn't matter as it just acts as a flag for nginx to return a 404

  *Note: this pattern is for rare use cases where a file should not be accessible from a particular site (ex: RW vs OCHA) but also demonstrates what could be done to accommodate some workflows or editorial policies*

**For private files**

The patterns are the same but instead of `public://media/` the base is `private://media/{provider-token}/` when `provider-token` is a hash specific to a provider.

*Note: the token is currently a hash of the provider api key and its uuid.*

To access private files, the client needs to pass the `X-Docstore-Provider-Uuid` and the `X-Docstore-Provider-Token` headers.

## File API endpoints

The media and file endpoints have been consolidated:

- `GET /api/v1/files` => Get files
- `POST /api/v1/files` => Create a file
- `GET /api/v1/files/{uuid}` => Get file info
- `PATCH /api/v1/files/{uuid}` => update filename or private state
- `DELETE /api/v1/files/{uuid}` => Delete file
- `GET /api/v1/files/{uuid}/content` => Get file content
- `POST /api/v1/files/{uuid}/content` => Create/update file content (creates a new revision if there was already content)
- `PUT /api/v1/files/{uuid}/select` => Select a revision for a provider (so permanent url point at it for provider)
- `PUT /api/v1/files/{uuid}/usage` => Get the list of resources referencing the file
- `GET /api/v1/files/{uuid}/revisions` => Get file revisions
- `GET /api/v1/files/{uuid}/revisions/{revision_id}` => Get a file revision
- `DELETE  /api/v1/files/{uuid}/revisions/{revision_id}` => Delete a revision (if not used by another provider)
- `GET /api/v1/files/{uuid}/revisions/{revision_id}/content` => Get a revision's content

## Tests

Existing tests have been modified to replace the media endpoints. Added also tests on the permanent URLs.

There is a `files` route as well, so this can be tested without nginx (ex: `docksal`).

A few more tests should be added about the `/api/v1/files/{uuid}/select` endpoint.

## Discussions

So with this approarch we have a client site proxying file requests (ex: `/attachments/{uuid}/{filename}` to the docstore `/files/{uuid}/{filename}` where nginx checks for a matching symlink and returns the content of the file. 

For private files (ex:`/private/attachments/{uuid}/{filename}`), the client verifies the access (user logged in for example) and then stream the content from the docstore `/files/{uuid}/{filename}` or proxy the request adding the `X-Docstore-Provider-Token` header. Like for public files, the docstore drupal site is never hit.

I think we could improve the performance even further (maybe that's a non issue) by mounting the docstore file directories in read-only so that the nginx of the client would directly look for the symlinks rather than proxying the request (basically copy the rules of the docstore nginx and mark the `/files/uuid/filename` location as internal).  @cafuego What do you think?

[OPS-7323]: https://humanitarian.atlassian.net/browse/OPS-7323

### CODE REVIEW

**Review by attiks — APPROVED** [review]
Looks good to me, feel free to merge

File: docker/etc/nginx/custom/01_docstore_files.conf [hunk 1]
```diff
@@ -0,0 +1,63 @@
+## Download handler.
+location ~ "^/files/([0-9a-f]{2})([0-9a-f]{2})([0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12})/([^/]+)(\.[0-9a-zA-Z]+)$" {
```
Comment (cafuego) [inline docker/etc/nginx/custom/01_docstore_files.conf:2]: I don't understand the `^/` match; we can't ever have files with a slash in their filename and the consumer should not be generating such URLs.

**Review by cafuego — APPROVED** [review]
I'm not at all convinced about embedding the token in directory paths to verify access; but at this stage I think it's better to just merge this PR and get *something* deployed so we can test it and work on it from there.

File: docker/etc/nginx/custom/01_docstore_files.conf (hunk 1 above)
Comment (orakili) [reply docker/etc/nginx/custom/01_docstore_files.conf:2]: @cafuego This is to make sure it only matches the pattern `/files/{uuid}/{name}.{ext}`. If we use something like `[^.]+` instead then `/files/1f4d4a9a-9013-40ed-a23b-445a3a76038e/thats/not/what/we/want.txt` for example would be a match.

**Pull request merged**